### ISSUE TITLE
Narrator focus is not moving in proper logical order when user is navigating by using caps + arrow keys (tab index is ignored)

### ISSUE THREAD

**Issue by M-Lipin** [dotnet/winforms#2164]
* .NET Core Version: 3.1
* Have you experienced this same bug with .NET Framework?: Yes

**Problem description:**
Narrator focus is not moving in proper logical order when user is navigating by using caps + arrow keys: Narrator goes thru the controls in order which is the order of controls being added to the Controls collection of the form in design time (order of controls can be corrected by manual changing the order of adding the control to the collection of child controls.

**Actual behavior:**
![image](https://user-images.githubusercontent.com/23213980/67387124-e3620300-f59e-11e9-99a5-a8668018df23.png)

**Expected behavior:**
Narrator caps + arrow navigation should take into account control tab index and accessible hierarchy should also take into account control tab index. 

**Minimal repro:**
* Create WinForms Application with one form
* Add TextBox1 to the bottom part of the form
* Add TextBox2 to the top part of form
* Add TextBox3 to the middle part of form
* Set tab index of TextBox2 to 1
* Set tab index of TextBox3 to 2
* Set tab index of TextBox1 to 3
* Run the app and navigate using Narrator.
* Observe Narrator caps + arrow right navigation goes firstly to TextBox1 -> TextBox2 -> TextBox3, but should go accordingly to tab index TextBox2 -> TextBox3 -> TextBox1

**Comment by M-Lipin** [comment]
**Workaround**: to add the controls to the Form.Controls in order accordingly to Tab index.

**Comment by weltkante** [comment]
I have played around with this suggestion and I don't think its applicable in general. WinForms designer appears to place controls in reverse order in the Form.Controls, the most recent control being at the front of the collection. When I look more closely the display order also is like this, the first control is on top and the last control is on bottom. For some reason ControlsCollection is in reverse order, so you need your top left control last and the bottom right first, sorting after TabIndex will screw things up I think.

Usually I place TabIndex is in text-flow order (left to right, top to bottom for me) but if you order controls after TabIndex they'll render in reverse (right to left, bottom to top) which will cause problems when things overlap - e.g. when text doesn't fit my autosized labels to the left of my textboxes are supposed to overlap under the textbox, not over it, which appears to be not possible anymore when sorting after TabIndex.

I'm a bit confused why the ControlsCollection is in reverse, never had payed much attention to it, but I'm sure there's a good technical reason for it ...

**Comment by M-Lipin** [comment]
Thank you for your comment @weltkante! It seems to me that minimal impact can be achieved by implementing custom accessible navigation for the form inner controls accordingly to tab index without changing the actual control order in the ControlsCollection.

**Comment by weltkante** [comment]
Yup its probably best to keep those concerns separated, the workaround described above is just that - a workaround - not something that can be done in general.

Besides the concerns with reversing order I described above, there are also issues when controls overlap - then you can't reorder anything at all because ControlCollection order is dictated by rendering order, and when things overlap you can't change that without breaking display results.

**Comment by RussKie** [comment]
> ControlsCollection is in reverse, never had payed much attention to it, but I'm sure there's a good technical reason for it ...

Z-order?
I'm sure @DustinCampbell can provide more insights.

**Comment by M-Lipin** [comment]
Z-order is applicable to the case when in the hierarchy of Form's control we have something like this:
```
Form
  Control 1
  Control 2
    SubControl 2-1
    SubControl 2-2
  Control 3
    SubControl 3-1
      SubSubControl 3-1-1
```
Do you mean this case? How the reversed order of Form controls [Control 3, Control 2, Control 1] can help with it? Sorry if I missed something.

**Comment by weltkante** [comment]
I think that was just a response to me wondering about the reverse ControlsCollection order

**Comment by weltkante** [comment]
@RussKie might have mistaken the PR comment?
> Prerequisite to fix #2164.

should not close this yet I think, nothing adressed this issue

**Comment by RussKie** [comment]
Aye, thank you. Looks like this is an epic issue :)
